modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/runner/TagHarvester.java

```diff
@@ -0,0 +1,118 @@
+
+    private IDynamicStatusStoreService dss;
+    private IResultArchiveStore ras;
+    private TestStructure testStructure;
+    private GalasaGson gson ;
+    public TagHarvester(IDynamicStatusStoreService dss, IResultArchiveStore ras, TestStructure testStructure, GalasaGson gson) {
+        this.dss = dss ;
+        this.ras = ras ;
+        this.testStructure = testStructure;
+        this.gson = gson;
+    }
+
+    /**
+     * We have a test class with possible annotations inside. 
+     * Harvest these and add them to the collection of tags we have from when the test was submitted.
+     * Then update the dss and ras accordingly.
+     * 
+     * @param testClass
+     */
+    public void harvestTagsFromTestClass( Class<?> testClass, String runName) {
+        Set<String> classAnnotationTags = extractTagsFromClass(testClass);
+        if( ! classAnnotationTags.isEmpty() ) {
+            // There were test class annotations which added tags.
+            updateTestStructureWithTags(classAnnotationTags);
+            updateTagsInDss(runName);
+            updateTagsInRas();
+        }
+    }
+
+    private Set<String> extractTagsFromClass(Class<?> clazz) {
+        HashSet<String> tags = new HashSet<String>();
+
+        Tags[] annotations = clazz.getAnnotationsByType(Tags.class);
+        if( annotations != null) {
+            for( Tags annotation : annotations ){
+                String[] values = annotation.value();
+                if( values != null) {
+                    for( String value : values) {
+                        if( value != null) {
+                            String trimmedValue = value.trim();
+                            if(!trimmedValue.isBlank() ){
+                                tags.add(trimmedValue);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return tags;
+    }
+
+    private void updateTestStructureWithTags(Set<String> tagsToAdd) {
+        // Update the test structure with tags.
+        for(String tag: tagsToAdd ) {
+            this.testStructure.addTag(tag);
+        }
+    }
+
+    private void updateTagsInDss(String runName) {
+
+        // Update the dss with the tags we have collected in the test structure.
+        Set<String> tags = this.testStructure.getTags();
+        String key = "run." + runName + "." + "tags" ;
+
+        // The tags are stored as a json array of strings... so 
+        // turn the tags into json.
+        JsonArray tagsJsonArray = new JsonArray();
+        for( String tag: tags) {
+            tagsJsonArray.add(tag);
+        }
+        String tagsJson = gson.toJson(tagsJsonArray);
+
+        try {
+            this.dss.put(key, tagsJson);
+        } catch (DynamicStatusStoreException ex) {
+            logger.error("Failed to write tags to dss. Ignoring and running the test anyway.",ex);
+        }
+    }
+
+    private void updateTagsInRas() {
```

Comment:
I was a bit worried that this is another write to couchdb... performance...etc. one more write for each test which has tags. It no tags then it won't get written though.

I could have left the 'dirty' TestStructure around until the next testcase state change to avoid this, but if it crashed/died before then you'd not have the tagging information collected in the failed run test structure in RAS...